org.eclipse.jdt.core.manipulation/proposals/org/eclipse/jdt/internal/ui/text/correction/proposals/TypeChangeCorrectionProposalCore.java

```diff
@@ -326,6 +386,64 @@ protected ASTRewrite getRewrite() throws CoreException {
 		return null;
 	}
 
+	private ITypeBinding[] getNewConstructorProposals() {
+		HashSet<CompletionProposal> completionProposals = new HashSet<>();
+		CompletionRequestor requestor = new CompletionRequestor(false) {
+			@Override
+			public void accept(CompletionProposal proposal) {
+				completionProposals.add(proposal);
+			}
+		};
+
+		requestor.setAllowsRequiredProposals(CompletionProposal.CONSTRUCTOR_INVOCATION, CompletionProposal.TYPE_REF, true);
+
+		try {
+			fCompilationUnit.codeComplete(fNode.getStartPosition() + SKIP_NEW_KEYWORD, requestor, new CompletionTimeoutProgressMonitor(1000));
+		} catch (OperationCanceledException e) {
+			// do nothing
+		} catch (JavaModelException e1) {
+			return null;
+		}
+		IJavaProject project= fCompilationUnit.getJavaProject();
+		List<String> typeSuggestions = new ArrayList<>();
+		for (CompletionProposal p: completionProposals) {
+			String typeName;
+			try {
+				typeName = p.getDeclarationSignature() != null ? new String(Signature.toCharArray(Signature.getTypeErasure(p.getDeclarationSignature()))) : null;
+			} catch (IllegalArgumentException e) {
+				typeName = p.getSignature() != null ? new String(Signature.toCharArray(Signature.getTypeErasure(p.getSignature()))) : null;
+			}
+			if (typeName != null && !TypeFilter.isFiltered(typeName) && !typeSuggestions.contains(typeName)) {
+				typeSuggestions.add(typeName);
+			}
+		}
+		IJavaElement[] typeElements= typeSuggestions.stream().map(b -> {
+			try {
+				return project.findType(b);
+			} catch (JavaModelException e) {
+				return null;
+			}
+		})
+				.filter(e -> {
+					try {
+						return e != null && !Modifier.isAbstract(e.getFlags());
+					} catch (JavaModelException e1) {
+						return false;
+					}
+				})
+				.toArray(IJavaElement[]::new);
+		ASTParser parser = ASTParser.newParser(IASTSharedValues.SHARED_AST_LEVEL);
+		parser.setProject(project);
+		ITypeBinding[] newBindings = Arrays.stream(parser.createBindings(typeElements, new NullProgressMonitor())).map(b -> (ITypeBinding)b)
+				.filter(b -> b.isAssignmentCompatible(((ITypeBinding) fBinding).getTypeDeclaration()) || TypeRules.canAssign(b, ((ITypeBinding) fBinding).getTypeDeclaration())
+						|| Arrays.stream(b.getInterfaces()).anyMatch(i -> ((ITypeBinding) fBinding).getBinaryName().equals(i.getBinaryName())))
```

Comment:
Instead of the interface logic you just added, could you try :

`Bindings.isSuperType(((ITypeBinding) fBinding).getTypeDeclaration(), tb, false)` ?